File: solutions/113.path-sum-ii/path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
const pathSum = (root, sum) => {
  if(!root) return []
  let res = []

  const dfs = (node, target, temp) => {

    if (node.left === null && node.right === null) {
      if (target === node.val) {
        res.push([...temp, node.val])
      }

      return
    }

    if (node.left) {
      dfs(node.left, target - node.val, [...temp, node.val])
    }

    if (node.right) {
      dfs(node.right, target - node.val, [...temp, node.val])
    }
  }

  dfs(root, sum, [])

  return res
}

File: solutions/113.path-sum-ii/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root:
            return []

        ret = []

        def traverse(node, temp, path):
            nonlocal ret
            # leaf
            if not node.left and not node.right:
                if temp == node.val:
                    ret.append(path + [node.val])
                    return

            if node.left:
                traverse(node.left, temp - node.val, path + [node.val])

            if node.right:
                traverse(node.right, temp - node.val, path + [node.val])

        traverse(root, sum, [])
        return ret
